app/src/main/java/com/github/libretube/ui/dialogs/DownloadPlaylistDialog.kt

```diff
@@ -0,0 +1,87 @@
+package com.github.libretube.ui.dialogs
+
+import android.app.Dialog
+import android.content.Intent
+import android.os.Bundle
+import androidx.core.content.ContextCompat
+import androidx.fragment.app.DialogFragment
+import com.github.libretube.R
+import com.github.libretube.constants.IntentData
+import com.github.libretube.databinding.DialogDownloadPlaylistBinding
+import com.github.libretube.enums.PlaylistType
+import com.github.libretube.extensions.getWhileDigit
+import com.github.libretube.extensions.serializable
+import com.github.libretube.helpers.LocaleHelper
+import com.github.libretube.services.PlaylistDownloadEnqueueService
+import com.google.android.material.dialog.MaterialAlertDialogBuilder
+
+class DownloadPlaylistDialog : DialogFragment() {
+    private lateinit var playlistId: String
+    private lateinit var playlistName: String
+    private lateinit var playlistType: PlaylistType
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        playlistId = requireArguments().getString(IntentData.playlistId)!!
+        playlistName = requireArguments().getString(IntentData.playlistName)!!
+        playlistType = requireArguments().serializable(IntentData.playlistType)!!
+    }
+
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+        val binding = DialogDownloadPlaylistBinding.inflate(layoutInflater)
+
+        val possibleVideoQualities = resources.getStringArray(R.array.defres).toList().let {
+            // remove the automatic quality entry
+            it.subList(1, it.size)
+        }
+        val possibleAudioQualities = resources.getStringArray(R.array.audioQualityBitrates)
+        val availableLanguages = LocaleHelper.getAvailableLocales()
+
+        binding.videoSpinner.items = listOf(getString(R.string.no_video)) + possibleVideoQualities
+        binding.audioSpinner.items = listOf(getString(R.string.no_audio)) + possibleAudioQualities
+        binding.subtitleSpinner.items =
+            listOf(getString(R.string.no_subtitle)) + availableLanguages.map { it.name }
+        binding.audioLanguageSpinner.items =
+            listOf(getString(R.string.default_language)) + availableLanguages.map { it.name }
+
+        return MaterialAlertDialogBuilder(requireContext())
+            .setTitle(getString(R.string.download_playlist) + ": " + playlistName)
+            .setMessage(R.string.download_playlist_note)
+            .setView(binding.root)
+            .setPositiveButton(R.string.download) { _, _ ->
+                val maxVideoQuality = binding.videoSpinner.selectedItemPosition.let { videoPosition ->
```

Comment:
Nitpicking (and debatable 😄): We might be able to make use of the `with` operator and List'`getOrNull` to possibly simplify this block:
```kotlin
    //...
    with (binding) {
        val videoQualityIndex = videoSpinner.selectedItemPosition
        //...

        if (videoQualityIndex == 0 && audioQualityIndex == 0) {
            return@setPositiveButton
        }

        val videoQuality = possibleVideoQualities.getOrNull(videoQualityIndex - 1)?.getWhileDigit()
        //...

        val downloadEnqueueIntent = Intent(requireContext(), PlaylistDownloadEnqueueService::class.java)
            .putExtra(IntentData.maxVideoQuality, videoQuality)
        //...
    }
```
Please, feel free to resolve.

Reply:
Yes, I thought about this too, but I ended up thinking that it's easier to read when we use an if/else block, because the `get(index-1)` is a bit like magic and it's probably not that easy to understand when looking at it again at some time.

Using `with` sounds useful though.

Reply:
Yep, I agree that the `getOrNull(index-1)` complexifies the block